multiraft/heartbeat_test.go

```diff
@@ -0,0 +1,152 @@
+// the function returns.
+func validateHeartbeatCount(ch <-chan *interceptMessage,
+	heartbeatCount, heartbeatResponseCount int, t *testing.T) {
+	// maps nodeID to heartbeats sent by this node minus heartbeat
+	// responses received by this node.
+	outInDelta := make(map[uint64]int)
+	cntH := 0
+	cntR := 0
+	defer func() {
+		for nodeID, delta := range outInDelta {
+			// If one of the checks below fails, this information will be useful.
+			log.Infof("node %v: %d heartbeats without response", nodeID, delta)
+		}
+		if cntH != heartbeatCount {
+			t.Errorf("unexpected total number of heartbeats: %d (expected %d)",
+				cntH, heartbeatCount)
+		}
+		if cntR != heartbeatResponseCount {
+			t.Errorf("unexpected total number of heartbeat responses: %d (expected %d)",
+				cntR, heartbeatResponseCount)
+		}
+	}()
+	for iMsg := range ch {
+		iMsg.ack <- struct{}{}
+		req := iMsg.args.(*RaftMessageRequest)
+		switch req.Message.Type {
+		case raftpb.MsgHeartbeat:
+			cntH++
+			outInDelta[req.Message.From]++
+		case raftpb.MsgHeartbeatResp:
+			cntR++
+			nodeID := req.Message.To
+			outInDelta[nodeID]--
+			if outInDelta[nodeID] < 0 {
+				t.Errorf("node %v: more responses received than heartbeats sent",
+					nodeID)
+			}
+		}
+		// If we've already received all the heartbeat responses that
+		// we have asked for, return. Without this, it's more awkward
+		// to know when to shut down the cluster.
+		if cntR >= heartbeatResponseCount {
+			return
+		}
+	}
+}
+
+func TestHeartbeat(t *testing.T) {
+	nodeCount := 3
+	cluster := newTestCluster(nodeCount, t)
+	// Outgoing messages block the client until we acknowledge them, and
+	// messages are sent synchronously.
+	cluster.transport.EnableEvents()
+
+	cluster.start()
+	leaderTickCount := 18
+	// Since waitForElection() currently triggers two ticks, it also triggers
+	// some heartbeats which remain unanswered (the sender isn't leader yet).
+	unansweredHeartbeats := (nodeCount - 1) * 2 // 2 because of 2 election ticks.
+
+	// How many heartbeats and heartbeat responses we expect in total.
+	expHeartbeatCount := (nodeCount-1)*leaderTickCount + unansweredHeartbeats
+	expHeartbeatRespCount := (nodeCount - 1) * leaderTickCount
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		// Run the actual checks, consuming from the Events channel.
+		validateHeartbeatCount(cluster.transport.Events, expHeartbeatCount,
+			expHeartbeatRespCount, t)
+		// Once we're here, we may begin tearing down.
+		wg.Done()
+		// For the rest of this test, simply keep acknowledging messages
+		// so that Raft makes progress. This mostly serves to avoid deadlock
+		// in cases where there is unexpected behaviour.
+		processEventsUntil(cluster.transport.Events, func(r *RaftMessageRequest) bool {
+			return false
+		})
+	}()
+
+	groupID := uint64(1)
+	leaderIndex := 0
+	cluster.createGroup(groupID, 3)
+	election := cluster.waitForElection(leaderIndex)
+
+	for i := 0; i < leaderTickCount; i++ {
```

Comment:
that's currently awkward because we intercept everything (where really we only want to block and acknowledge heartbeats) but I will keep it in mind for the interceptingLocaTransport.